ecmascript/transforms/base/src/resolver/tests.rs

```diff
@@ -542,11 +539,11 @@ to!(
         }(Bar);
     "#,
     r#"
-    var Foo = function(_Bar) {
-            _inherits(Foo, _Bar);
-            function Foo() {
+    var Foo1 = function(_Bar) {
```

Comment:
Changes like this are due to an issue in the hygiene pass that already existed: if a variable is referenced before it is defined, it will be renamed. For example:

```js
var Foo;
function test() {
  test(Foo);
  function Foo() {}
}
```

is compiled to:

```js
var Foo1;
function test() {
    test(Foo2);
    function Foo2() {
    }
}
```

That's due to this:

https://github.com/swc-project/swc/blob/8f5daa3bbbcc05997a71dd23159347ad3622238c/ecmascript/transforms/base/src/hygiene/mod.rs#L116-L118

The code works but more vars are renamed than really needed. Removing those lines fixes this issue, but breaks some other tests which seem to rely on this behavior so I didn't do it. Perhaps could be tackled separately?

Reply:
```js
var Foo1;
function test() {
    test(Foo2);
    function Foo2() {
    }
}
```


Isn't this correct? I thought the function `Foo` is hoisted, so `Foo` in `test` refers to the function, not a variable from the parent scope.


If this is wrong, then the implementation is wrong and should be fixed.

Reply:
Yeah the output is "correct", but the outer `Foo` was unnecessarily renamed to `Foo1`. If you reorder the statements such that the call to `test` is after the declaration, it is not renamed.

Reply:
I See.
Then it seems like an issue that should be fixed separately.
Thanks for the clarification!.